UP008 false negative in inner (nested) classes

### Summary

Playground link: https://play.ruff.rs/f9f8f76b-c468-4f4f-a3b8-0bca5850daf1

```python
class Base:
    def __init__(self, foo):
        self.foo = foo

class Outer(Base):
    def __init__(self, foo):
        super(Outer, self).__init__(foo)  # UP008 issued as expected

    class Inner(Base):
        def __init__(self, foo):
            super(Outer.Inner, self).__init__(foo)  # Bug: would also expect UP008, but it was not issued

o = Outer(5)
i = Outer.Inner(5)
```

---

By the way, Pylint's [`super-with-arguments`](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/super-with-arguments.html) also has a problem only with `super(Outer, self)`, but not with `super(Outer.Inner, self)`:

```console
$ pylint --version
pylint 4.0.4
astroid 4.0.3
Python 3.14.2 (tags/v3.14.2:df79316, Dec  5 2025, 17:18:21) [MSC v.1944 64 bit (AMD64)]
$ pylint --disable=all --enable=super-with-arguments test.py
************* Module test
test.py:7:8: R1725: Consider using Python 3 style super() without arguments (super-with-arguments)

-----------------------------------
Your code has been rated at 9.09/10
```

However, https://github.com/asottile/pyupgrade correctly upgrades both `super()` calls:

```console
$ pip freeze | grep -F pyupgrade
pyupgrade==3.21.2
$ cp test.py test-original.py
$ pyupgrade test.py
Rewriting test.py
$ git diff --no-index test-original.py test.py
diff --git 1/test-original.py 2/test.py
index 2f6d7cd2..65bf24ac 100644
--- 1/test-original.py
+++ 2/test.py
@@ -4,11 +4,11 @@ class Base:
 
 class Outer(Base):
     def __init__(self, foo):
-        super(Outer, self).__init__(foo)  # UP008 issued as expected
+        super().__init__(foo)  # UP008 issued as expected
 
     class Inner(Base):
         def __init__(self, foo):
-            super(Outer.Inner, self).__init__(foo)  # Bug: would also expect UP008, but it was not issued
+            super().__init__(foo)  # Bug: would also expect UP008, but it was not issued
 
 o = Outer(5)
 i = Outer.Inner(5)
```

### Version

ruff 0.14.11

--- a/crates/ruff_linter/resources/test/fixtures/pyupgrade/UP008.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyupgrade/UP008.py
@@ -315,5 +315,31 @@ class Outer(Base):
 
     class Inner(Base):
         def __init__(self, foo):
-            super(Outer.Inner, self).__init__(foo)  # Should trigger UP008
+            super(Outer.Inner, self).__init__(foo)  # UP008: matches enclosing class chain
+
+
+# super() first arg is an attribute that only matches on the last segment,
+# but refers to a different class (Outer.Inner, not __class__ which is Inner).
+class Inner(Outer.Inner):
+    def __init__(self, foo):
+        super(Outer.Inner, self).__init__(foo)  # Should NOT trigger UP008
+
+# super() first arg has an unrelated module prefix
+class Outer2:
+    class Inner2(Base):
+        def __init__(self, foo):
+            super(some_module.Outer2.Inner2, self).__init__(foo)  # Should NOT trigger UP008
+
+# 3-level deep nesting: super(A.B.C, self) should trigger UP008
+class A:
+    class B:
+        class C(Base):
+            def __init__(self, foo):
+                super(A.B.C, self).__init__(foo)  # UP008: matches full chain
+
+# Mismatched middle segment: Wrong.Inner doesn't match Outer3.Inner
+class Outer3:
+    class Inner(Base):
+        def __init__(self, foo):
+            super(Wrong.Inner, self).__init__(foo)  # Should NOT trigger UP008
 

--- a/crates/ruff_linter/src/rules/pyupgrade/rules/super_call_with_parameters.rs
+++ b/crates/ruff_linter/src/rules/pyupgrade/rules/super_call_with_parameters.rs
@@ -122,14 +122,6 @@ pub(crate) fn super_call_with_parameters(checker: &Checker, call: &ast::ExprCall
         return;
     };
 
-    let first_arg_id = match first_arg {
-        // Simple case: super(MyClass, self)
-        Expr::Name(ast::ExprName { id, .. }) => id,
-        // Nested class case: super(OuterClass.InnerClass, self)
-        Expr::Attribute(ast::ExprAttribute { attr, .. }) => &attr.id,
-        _ => return,
-    };
-
     let Expr::Name(ast::ExprName {
         id: second_arg_id, ..
     }) = second_arg
@@ -137,14 +129,42 @@ pub(crate) fn super_call_with_parameters(checker: &Checker, call: &ast::ExprCall
         return;
     };
 
-    // The `super(__class__, self)` and `super(ParentClass, self)` patterns are redundant in Python 3
-    // when the first argument refers to the implicit `__class__` cell or to the enclosing class.
-    // Avoid triggering if a local variable shadows either name.
-    if !(((first_arg_id == "__class__") || (first_arg_id == parent_name.as_str()))
-        && !checker.semantic().current_scope().has(first_arg_id)
-        && second_arg_id == parent_arg.name().as_str())
-    {
+    if second_arg_id != parent_arg.name().as_str() {
         return;
+    }
+
+    // Verify the first argument matches the enclosing class chain.
+    // For `super(__class__, self)` or `super(ClassName, self)`, just check the immediate class.
+    // For `super(Outer.Inner, self)`, verify each segment matches the enclosing class nesting.
+    match first_arg {
+        Expr::Name(ast::ExprName { id, .. }) => {
+            if !((id == "__class__" || id == parent_name.as_str())
+                && !checker.semantic().current_scope().has(id))
+            {
+                return;
+            }
+        }
+        Expr::Attribute(_) => {
+            let chain = collect_attribute_chain(first_arg);
+            // The innermost name must match the immediately enclosing class.
+            if chain.last() != Some(&parent_name.as_str()) {
+                return;
+            }
+            // Each preceding name must match the next enclosing class.
+            for name in chain.iter().rev().skip(1) {
+                let Some(Stmt::ClassDef(ast::StmtClassDef {
+                    name: enclosing_name,
+                    ..
+                })) = parents.find(|stmt| stmt.is_class_def_stmt())
+                else {
+                    return;
+                };
+                if *name != enclosing_name.as_str() {
+                    return;
+                }
+            }
+        }
+        _ => return,
     }
 
     drop(parents);
@@ -197,6 +217,29 @@ pub(crate) fn super_call_with_parameters(checker: &Checker, call: &ast::ExprCall
             applicability,
         ));
     }
+}
+
+/// Collects the chain of names from an attribute expression.
+///
+/// For example, `A.B.C` returns `["A", "B", "C"]`.
+fn collect_attribute_chain(expr: &Expr) -> Vec<&str> {
+    let mut chain = Vec::new();
+    let mut current = expr;
+    loop {
+        match current {
+            Expr::Attribute(ast::ExprAttribute { value, attr, .. }) => {
+                chain.push(attr.id.as_str());
+                current = value;
+            }
+            Expr::Name(ast::ExprName { id, .. }) => {
+                chain.push(id.as_str());
+                break;
+            }
+            _ => return Vec::new(),
+        }
+    }
+    chain.reverse();
+    chain
 }
 
 /// Returns `true` if a call is an argumented `super` invocation.

